### PR TITLE
Birthday feature

### PR DESCRIPTION
**Pull request by TerryAg**
Users can now add their Birthday via PCSocBot to automatically be granted the Birthday! role on their birthday. The role will last until the day is up.

Type `!birthday add <dd/mm>` to add your birthday to the queue, or type `!birthday remove` if you no longer want PCSocBot to track it.

### CODE REVIEW

File: commands/birthday.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import asyncio
+from collections import defaultdict
+import datetime
+import json
+import re
```
Comment (VineBalloon) [inline commands/birthday.py]: Unused import?

File: commands/birthday.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+from commands.base import Command
+from helpers import CommandFailure
+
+BIRTHDAY_FILE = "files/birthdays.json"
+BIRTHDAY_ROLE_NAME = "Birthday!"
+
```
Comment (VineBalloon) [inline commands/birthday.py]: PEP8 requires 2 newlines after imports/constants if you wish to follow it.

File: commands/birthday.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    try:
+        with open(bday_file) as birthdays:
+            all_birthdays.update(json.load(birthdays))
+    except FileNotFoundError:
+        pass
+
+    return all_birthdays
+
+
+def validate(date_string):
```
Comment (VineBalloon) [inline commands/birthday.py]: I see what you did there

File: commands/birthday.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        new = datetime.datetime.today()
+        if new.day != prev.day:
+            # It's a new day - remove all previous roles, add new roles
+            all_birthdays = get_birthdays(BIRTHDAY_FILE)
+            dm_today = new.strftime("%d/%m")
+            
+            # Get all members
+            server = list(client.servers)[0]
+            members = server.members
+            birthday_role = next(x for x in server.roles if x.name == BIRTHDAY_ROLE_NAME)
```
Comment (VineBalloon) [inline commands/birthday.py]: Consider using the [discord find util](https://discordpy.readthedocs.io/en/latest/api.html#discord.utils.find)

File: commands/birthday.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+            # Remove everyone with the Birthday role from yesterday
+            for member in members:
+                if any(birthday_role == role for role in member.roles):
+                    await client.remove_roles(member, birthday_role)
+
+            # Happy Birthday!
+            for birthday_member in all_birthdays[dm_today]:
+                member = server.get_member(birthday_member)
+                if member is not None:
```
Comment (VineBalloon) [inline commands/birthday.py]: You can do `if member:` and that checks for None.

File: commands/birthday.py (hunk 1 above)
Comment (TerryAg) [reply commands/birthday.py]: Good catch thanks!

File: commands/birthday.py (hunk 2 above)
Comment (TerryAg) [reply commands/birthday.py]: Done

File: commands/birthday.py (hunk 4 above)
Comment (TerryAg) [reply commands/birthday.py]: Done, ty

File: commands/birthday.py (hunk 5 above)
Comment (TerryAg) [reply commands/birthday.py]: Resolved in discord but yea nah I disagree per [PEP8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) (second bullet point).

**Pull request merged**